crates/e2e/tests/e2e/cow_amm.rs

```diff
@@ -0,0 +1,348 @@
+use {
+    app_data::AppDataHash,
+    e2e::{
+        setup::{colocation::SolverEngine, *},
+        tx,
+        tx_value,
+    },
+    ethcontract::{web3::ethabi::Token, BlockId, BlockNumber, U256},
+    model::{
+        order::{OrderCreation, OrderData, OrderKind},
+        signature::{hashed_eip712_message, EcdsaSigningScheme},
+    },
+    secp256k1::SecretKey,
+    shared::ethrpc::Web3,
+    solvers_dto::solution::{
+        BuyTokenBalance,
+        Call,
+        Kind,
+        SellTokenBalance,
+        SigningScheme,
+        Solution,
+    },
+    std::collections::HashMap,
+    web3::signing::SecretKeyRef,
+};
+
+#[tokio::test]
+#[ignore]
+async fn local_node_cow_amm() {
+    run_test(cow_amm).await;
+}
+
+async fn cow_amm(web3: Web3) {
+    let mut onchain = OnchainComponents::deploy(web3.clone()).await;
+
+    let [solver] = onchain.make_solvers(to_wei(100)).await;
+    let [bob, cow_amm_owner] = onchain.make_accounts(to_wei(1000)).await;
+
+    let [dai] = onchain
+        .deploy_tokens_with_weth_uni_v2_pools(to_wei(300_000), to_wei(100))
+        .await;
+
+    // Fund the buffers with a lot of buy tokens so we can pay out the required
+    // tokens for 2 orders in the same direction without having to worry about
+    // getting the liquidity on-chain.
+    dai.mint(onchain.contracts().gp_settlement.address(), to_wei(100_000))
+        .await;
+
+    // set up cow_amm
+    let oracle = contracts::CowAmmUniswapV2PriceOracle::builder(&web3)
+        .deploy()
+        .await
+        .unwrap();
+
+    let cow_amm_factory = contracts::CowAmmConstantProductFactory::builder(
+        &web3,
+        onchain.contracts().gp_settlement.address(),
+    )
+    .deploy()
+    .await
+    .unwrap();
+
+    // Fund cow amm owner with 2_000 dai and allow factory take them
+    dai.mint(cow_amm_owner.address(), to_wei(2_000)).await;
+    tx!(
+        cow_amm_owner.account(),
+        dai.approve(cow_amm_factory.address(), to_wei(2_000))
+    );
+    // Fund cow amm owner with 1 WETH and allow factory take them
+    tx_value!(
+        cow_amm_owner.account(),
+        to_wei(1),
+        onchain.contracts().weth.deposit()
+    );
+    tx!(
+        cow_amm_owner.account(),
+        onchain
+            .contracts()
+            .weth
+            .approve(cow_amm_factory.address(), to_wei(1))
+    );
+
+    let pair = onchain
+        .contracts()
+        .uniswap_v2_factory
+        .get_pair(onchain.contracts().weth.address(), dai.address())
+        .call()
+        .await
+        .expect("failed to get Uniswap V2 pair");
+
+    let cow_amm = cow_amm_factory
```

Comment:
nit: I believe `factory.create` returns the address of the created contract so no need to compute it like this.

Reply:
Yeah, it does. But to actually create the amm on-chain we have to use `.send()` instead of `.call()` and `.send()` only return the tx receipt. And AFAIK there is no good way to get the return value out of that.
So either we use `.create()` once with `.call()` to get the address and once with `.send()` to actually create the amm or we use the way I have currently.
Of course if you know how to nicely get the return value from `.send()` we can do that as well.